File: core-examples/src/main/java/io/vertx/example/core/http/proxyconnect/Proxy.java
package io.vertx.example.core.http.proxyconnect;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Launcher;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.net.NetClient;
import io.vertx.core.net.NetClientOptions;
import io.vertx.core.net.NetSocket;

/*
 * @author <a href="http://tfox.org">Tim Fox</a>
 */
public class Proxy extends AbstractVerticle {

  public static void main(String[] args) {
    Launcher.executeCommand("run", Proxy.class.getName());
  }

  @Override
  public void start() throws Exception {

    NetClient client = vertx.createNetClient(new NetClientOptions());

    vertx.createHttpServer().requestHandler(req -> {
      if (req.method() == HttpMethod.CONNECT) {

        // Determine proxied server address
        String proxyAddress = req.uri();
        int idx = proxyAddress.indexOf(':');
        String host = proxyAddress.substring(0, idx);
        int port = Integer.parseInt(proxyAddress.substring(idx + 1));

        System.out.println("Connecting to proxy " + proxyAddress);
        client.connect(port, host, ar -> {

          if (ar.succeeded()) {
            System.out.println("Connected to proxy");

            NetSocket serverSocket = ar.result();
            serverSocket.pause();

            req.toNetSocket().onComplete(ar2 -> {
              if (ar2.succeeded()) {
                NetSocket clientSocket = ar2.result();
                serverSocket.handler(buff -> {
                  System.out.println("Forwarding server packet to the client");
                  clientSocket.write(buff);
                });
                serverSocket.closeHandler(v -> {
                  System.out.println("Server socket closed");
                  clientSocket.close();
                });
                clientSocket.handler(buff -> {
                  System.out.println("Forwarding client packet to the server");
                  serverSocket.write(buff);
                });
                clientSocket.closeHandler(v -> {
                  System.out.println("Client socket closed");
                  serverSocket.close();
                });
                serverSocket.resume();
              } else {
                serverSocket.close();
              }
            });
          } else {

            System.out.println("Fail proxy connection");
            req.response().setStatusCode(403).end();
          }
        });
      } else {
        req.response().setStatusCode(405).end();
      }
    }).listen(8080);
  }
}

File: core-examples/src/main/kotlin/io/vertx/example/core/http/proxyconnect/Proxy.kt
package io.vertx.example.core.http.proxyconnect

import io.vertx.core.http.HttpMethod
import io.vertx.core.net.NetClientOptions
import io.vertx.kotlin.core.net.*

class Proxy : io.vertx.core.AbstractVerticle()  {
  override fun start() {

    var client = vertx.createNetClient(NetClientOptions())

    vertx.createHttpServer().requestHandler({ req ->
      if (req.method() == HttpMethod.CONNECT) {

        // Determine proxied server address
        var proxyAddress = req.uri()
        var idx = proxyAddress.indexOf(':')
        var host = proxyAddress.substring(0, idx)
        var port = java.lang.Integer.parseInt(proxyAddress.substring(idx + 1))

        println("Connecting to proxy ${proxyAddress}")
        client.connect(port, host, { ar ->

          if (ar.succeeded()) {
            println("Connected to proxy")
            var clientSocket = req.netSocket()
            clientSocket.write("HTTP/1.0 200 Connection established\n\n")
            var serverSocket = ar.result()

            serverSocket.handler({ buff ->
              println("Forwarding server packet to the client")
              clientSocket.write(buff)
            })
            serverSocket.closeHandler({ v ->
              println("Server socket closed")
              clientSocket.close()
            })

            clientSocket.handler({ buff ->
              println("Forwarding client packet to the server")
              serverSocket.write(buff)
            })
            clientSocket.closeHandler({ v ->
              println("Client socket closed")
              serverSocket.close()
            })
          } else {

            println("Fail proxy connection")
            req.response().setStatusCode(403).end()
          }
        })
      } else {
        req.response().setStatusCode(405).end()
      }
    }).listen(8080)
  }
}
